### ISSUE TITLE
Add parameter to pass initial phase from the Python side

### ISSUE THREAD

**Issue by SimeonEhrig** [ComputationalRadiationPhysics/student_project_python_bindings#2]
To verify that the result of the reconstruction does not depend on the random number generator of the initial phase, it should be possible that the initial phase is generated with the Numpy random number generator and passed to the CUDA C++ code.

Should be done after issue #1 

# Possible implementation

```C++
py::array_t<double> fienup_phase_retrieval(py::array_t<complex<double>> mag, py::array_t<double> masks, int steps, bool verbose, string mode, double beta, py::array_t<complex<double>> * initial_phase = nullptr)
{
  // ...
  if(initial_phase){
      cudaMemcpy(y_hat_dev, initial_phase, sizeof(initial_phase), cudaMemcpyHostToDevice);
  } else {
      //sample random phase
      get_yhat_dev1<<<size_y, size_x>>>(states, y_hat_dev, mag_dev, size_y, size_x)
  }
  //...
}
```

**Comment by SimeonEhrig** [comment]
We start a small conversation on a side channel. @afif-ishamsyah asked:
1. Do I need to replace `curand_uniform(&states[idx])` to numpy random?
2. I think this is also related to issue #7, because I don't use any C standard random library?

- Answer to question 1.: No, `curand_uniform` should be not replaced. The numpy random generator should be usable as alternative for testing. This means, if there is no initial Phase passed from the Python side, use `curand_uniform` to generate one.  We should keep `curand_uniform` because it is more performant.

- Answer to question 2: Is the [srand()](https://github.com/ComputationalRadiationPhysics/student_project_python_bindings/blob/0747ef314f912253cb25d7d5e0ffba603328ad55/my_phase-retrieval/phase_algo.hpp#L37) used for the curand library?

**Comment by SimeonEhrig** [comment]
No, other. I explain it from the Python side.

```python
import cuPhaseRet #c++ pybind phase retrieval
import numpy as np
import imageio
import matplotlib.pyplot as plt
# from phase_retrieval import fienup_phase_retrieval
from time import perf_counter 

np.random.seed(1)
image = imageio.imread('b.png', as_gray=True)
magnitudes = np.abs(np.fft.fft2(image))

# fineup_phase_retrievial uses internal curand to initialize y_hat_dev
result =  cuPhaseRet.fienup_phase_retrieval(magnitudes, mask, 500, False, "hybrid", 0.8)
```

```python
import cuPhaseRet #c++ pybind phase retrieval
import numpy as np
import imageio
import matplotlib.pyplot as plt
# from phase_retrieval import fienup_phase_retrieval
from time import perf_counter 

np.random.seed(1)
image = imageio.imread('b.png', as_gray=True)
magnitudes = np.abs(np.fft.fft2(image))

# create initial phase
y_hat = mag*np.exp(1j*2*np.pi*np.random.rand(*mag.shape))

# fineup_phase_retrievial copies y_hat to GPU (y_hat_dev) instead using curand
result =  cuPhaseRet.fienup_phase_retrieval(magnitudes, mask, 500, False, "hybrid", 0.8, y_hat)
```

**Comment by afif-ishamsyah** [comment]
okay. so there is 3 version of it. CURAND, numpy random, and c++ random library

**Comment by SimeonEhrig** [comment]
I believe only 2 or do you use the c or c++ random library at the moment? If not, you can skip it and it was a mistake from my side, if understand that you use it.

**Comment by afif-ishamsyah** [comment]
For the original random_phase, I only use curand. In the latest commit today, I create a second version of phase retrieval that use c++ random library. I will create another which is using numpy random next time.

**Comment by afif-ishamsyah** [comment]
It is confirmed that with the same random value and the same image, the result of python and CUDA version is the same. Is this solved?

**Comment by SimeonEhrig** [comment]
Yes, solved with commit fd7be4e12656390426931838bacbfda05b321746
